#! /bin/sh /usr/share/dpatch/dpatch-run
## 001_WorkingDir.dpatch by  <SvenDowideit@home.org.au>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: reduce the security footprint of working/tmp

@DPATCH@

diff -urNad /tmp/build_deb/twiki-4.1.2/lib/TWiki.pm TWiki412/lib/TWiki.pm 
--- /tmp/build_deb/twiki-4.1.2/lib/TWiki.pm	2007-03-04 01:45:57.000000000 +1100
+++ TWiki412/lib/TWiki.pm	2007-10-29 13:47:09.000000000 +1100
@@ -787,7 +787,9 @@
         }
     }
 
-    open(F, ">$passthruFilename") || die "{TempfileDir} cache not writable $!";
+    use Fcntl;
+    #passthrough file is only written to once, so if it already exists, suspect a security hack (O_EXCL)
+    sysopen(F, "$passthruFilename", O_RDWR|O_EXCL|O_CREAT, 0644) || die "{TempfileDir} cache not writable $!";
     $query->save(\*F);
     close(F);
     return 'twiki_redirect_cache='.$uid;

diff -urNad /tmp/build_deb/twiki-4.1.2/lib/TWiki/Client.pm TWiki412/lib/TWiki/Client.pm 
--- /tmp/build_deb/twiki-4.1.2/lib/TWiki/Client.pm	2007-03-04 01:45:57.000000000 +1100
+++ TWiki412/lib/TWiki/Client.pm	2007-10-29 13:25:03.000000000 +1100
@@ -181,7 +181,7 @@
     return undef unless $ip; # no IP address, can't map
 
     my %ips;
-    if( open( IPMAP, '<', $TWiki::cfg{TempfileDir}.'/ip2sid' )) {
+    if( open( IPMAP, '<', $TWiki::cfg{WorkingDir}.'/ip2sid' )) {
         local $/ = undef;
         %ips = map { split( /:/, $_ ) } split( /\r?\n/, <IPMAP> );
         close(IPMAP);
@@ -189,7 +189,7 @@
     if( $sid ) {
         # known SID, map the IP addr to it
         $ips{$ip} = $sid;
-        open( IPMAP, '>', $TWiki::cfg{TempfileDir}.'/ip2sid') ||
+        open( IPMAP, '>', $TWiki::cfg{WorkingDir}.'/ip2sid') ||
           die "Failed to open ip2sid map for write. Ask your administrator to make sure that the {Sessions}{Dir} is writable by the webserver user.";
         print IPMAP map { "$_:$ips{$_}\n" } keys %ips;
         close(IPMAP);
